base/loading.jl

```diff
@@ -1093,6 +1094,16 @@ function _require_prelocked(uuidkey::PkgId)
         error("package `$(uuidkey.name)` did not define the expected \
               module `$(uuidkey.name)`, check for typos in package module name")
     end
+    pkgorig = get(pkgorigins, uuidkey, nothing)
+    new_path = locate_package(uuidkey)
+    if pkgorig !== nothing && pkgorig.path !== nothing &&
+            !samefile(fixup_stdlib_path(pkgorig.path), new_path) && uuidkey.uuid ∉ already_warned_path_change_pkgs
+        @warn "Package $(uuidkey.name) already loaded from path $(repr(pkgorig.path)), \
+               now attempted to be loaded from $(repr(new_path)). This might indicate a \
+               change of environment between package loads and can mean that incompatible \
+               packages have been loaded."
```

Comment:
I think showing versions would be helpful.
Also I've added a shorter startup.jl mention than my previous suggestion.

```suggestion
        cur_vstr = isnothing(pkgorig.version) ? "" : string(" (", pkgorig.version, ")")
        new_v = get_pkgversion_from_path(new_path)
        new_vstr = isnothing(new_v) ? "" : string(" (", new_v, ")")
        warnstr = "Package $(uuidkey.name)$cur_vstr already loaded from path $(repr(pkgorig.path)), \
            now attempted to be loaded from $(repr(new_path))$new_vstr. This might indicate a \
            change of environment between package loads and can mean that incompatible \
            packages have been loaded"
        if Base.JLOptions.startupfile == 1
                warnstr *= ", perhaps from a startup.jl. Consider starting directly in the project \
                    via the `--project` arg."
        else
                warnstr *= "."
        end
        @warn warnstr
```

Reply:
Bump @KristofferC 

I don't mind taking this over if you're ok with that

Reply:
Bump

Reply:
> I don't mind taking this over if you're ok with that

Sure, please go ahead.